File: 백트래킹/9663/9663.c
#include <stdio.h>
#define MAX 15

int find(int, int);
int nqueen(int);
int arr[MAX];
int l;

int main()
{
   scanf("%d", &l);
   printf("%d", nqueen(0));

   return 0;
}

int find(int n, int val)
{
   int i = 0;
   for (i = 0; i < n; ++i)
   {
      if (arr[i] == val)
         return i;
      if (arr[i] - i == val - n)
         return i;
      if (arr[i] + i == val + n)
         return i;
   }
   return -1;
}

int nqueen(int n)
{
   int i, answer;
   if (l == n)
      return 1;
   answer = 0;
   for (i = 0; i < l; ++i)
   {
      if (find(n, i) != -1)
         continue;
      arr[n] = i;
      answer += nqueen(n + 1);
   }
   return answer;
}

File: 백트래킹/9663/9663.py
def find(arr, val):
   for i in range(len(arr)):
      if arr[i] == val:
         return i
      if arr[i] - i == val - len(arr):
         return i
      if arr[i] + i == val + len(arr):
         return i
   return -1

def nqueen(arr, n):
   if (len(arr) == n):
      return 1
   answer = 0
   for i in range(n):
      if find(arr, i) != -1:
         continue
      temp = arr.copy()
      temp.append(i)
      answer += nqueen(temp, n)
   return answer
  
n = int(input())
print(nqueen([], n))